tools/clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5097,14 +5097,65 @@ static void ProcessDeclAttribute(Sema &S, Scope *scope, Decl *D,
   }
 }
 
+void Sema::ValidateShaderAttribtues(Scope *S, Decl *D, const AttributeList *A) {
+  bool hasComputeAttr = false;
+  bool hasNodeAttr = false;
+  int attrNum = 0;
+
+  for (const AttributeList *l = A; l; l = l->getNext()) {
+    if (l->getKind() == AttributeList::AT_HLSLShader) {
+      Expr *E = l->getArgAsExpr(0);
+      if (E->isTypeDependent() || E->isValueDependent() ||
+          E->getStmtClass() != Stmt::StringLiteralClass) {
+        break;
+      }
+
+      StringLiteral *sl = cast<StringLiteral>(E);
+      std::string str = sl->getString();
+      if (str == "compute") {
+        hasComputeAttr = true;
+        attrNum++;
+      } else if (str == "node") {
+        hasNodeAttr = true;
+        attrNum++;
+      } else {
+        for (auto shaderAttrStr :
+             {"vertex", "pixel", "hull", "domain", "geometry", "raygeneration",
+              "intersection", "anyhit", "closesthit", "miss", "callable",
+              "mesh", "amplification"}) {
```

Comment:
InB4: Is there a way to compare these without using strings that wouldn't involve a major reworking of how the attribs are stored? I suspect the answer is no, but I think it needs to be stated.